aws_lambda_builders/workflows/dotnet_clipackage/DESIGN.md

```diff
@@ -0,0 +1,86 @@
+This tool has several commands for building and publishing .NET Core Lambda functions. For this integration 
+the `dotnet lambda package` command will be used to create a zip file that can be deployed to Lambda.
+
+The builder will install the Amazon.Lambda.Tools Global Tool or update to the latest version before executing 
+the package command.
+
+This builder assumes the [.NET Core command-line interface (CLI)](https://docs.microsoft.com/en-us/dotnet/core/tools/?tabs=netcore2x) 
+is already installed and added to the path environment variable. This is a reasonable requirement as the 
+.NET Core CLI is a required tool for .NET Core developers to build any .NET Core project.
+
+The .NET Core CLI handles the validation that the correct version of .NET Core is installed and errors out when there is 
+not a correct version. 
+
+### Challenges
+
+#### Output
+
+The output of `dotnet lambda package` command is a zip archive that consumers can then deploy to Lambda. For SAM build
+the expected output is a directory of all of the output files. To make the package command compatible with the SAM build 
+this builder will direct the package command to output the zip file in the artifacts folder. Once the package command is complete
+it expands the zip file and then deletes the zip file.
+
+#### Parameters
+
+The package command takes in serveral parameters. Here is the help for the package command.
+```bash
+> dotnet lambda package --help
+Amazon Lambda Tools for .NET Core applications (3.1.2)
+Project Home: https://github.com/aws/aws-extensions-for-dotnet-cli, https://github.com/aws/aws-lambda-dotnet
+
+package:
+   Command to package a Lambda project into a zip file ready for deployment
+
+   dotnet lambda package [arguments] [options]
+   Arguments:
+      <ZIP-FILE> The name of the zip file to package the project into
+   Options:
+      -c     | --configuration                Configuration to build with, for example Release or Debug.
+      -f     | --framework                    Target framework to compile, for example netcoreapp2.1.
+      --msbuild-parameters                    Additional msbuild parameters passed to the 'dotnet publish' command. Add quotes around the value if the value contains spaces.
+      -pl    | --project-location             The location of the project, if not set the current directory will be assumed.
+      -cfg   | --config-file                  Configuration file storing default values for command line arguments.
+      -pcfg  | --persist-config-file          If true the arguments used for a successful deployment are persisted to a config file.
+      -o     | --output-package               The output zip file name
+      -dvc   | --disable-version-check        Disable the .NET Core version check. Only for advanced usage.
+```
+
+Currently **--framework** is the only required parameter which tells the underlying build process what version of .NET Core to build for.
```

Comment:
How is this infomation passed in? is it decided by the builder based on the runtime?

Reply:
For most users --framework will be already set in the aws-lambda-tools-defaults.json file so they won't have to do anything. If it is not set then the user would need to specify it on the sam build command line which would get passed into the builder. I tried to follow the examples I saw in the other builders for grabbing options that were passed on the commandline.

Reply:
So this is currently not supported in SAM CLI, although the correct plumbing exists thru `options` property. See https://github.com/awslabs/aws-lambda-builders/issues/4 

How big of a requirement is this? Do you know the fraction of users that use parameters vs .json file to pass this value?

Reply:
It is not big requirement since the framework will most likely be set through the *.json file. It is a good thing to have for the future. As long as the code is doing it in the way SAM envisions this being handled in the future is all that I require now. We should make sure SAM init version of a .NET Core project includes the json file like all of the rest of our .NET tooling does.

I also might update the Amazon.Lambda.Tools to auto detect the framework based on the project file. If/when I do this work SAM build would pull that change in automatically due to the nature of this PR always updating the version of Amazon.Lambda.Tools.

Reply:
Yeah, in the future when we add add `sam build --debug` mode, we would want to pass this over to lambda-builders, and the dotnet workflow to pick it up as the configuration.

This would also yield a conversation on which is higher priority, the flag or the information in the json config file.

Reply:
The existing Amazon.Lambda.Tools .NET Core Global tool that this wraps around already has the rule that if a value is specified in the command line it overrides any setting in the json file. I would expect that rule to apply from sam build as well.